### DISCUSSION TITLE
Is there a way to have workspace for `Edk2PlatformSetup().Invoke()` within repo subfolder?

### DISCUSSION THREAD

**Discussion by ljanik** [tianocore/edk2-pytool-extensions#585] [Q&A]
Hello,
on our embedded project we integrate EDK as submodules to main project repository. Those submodules are however not in the root of the main repo - they are concentrated in a subfolder, see below:

```
doc
atf
mkimage
optee-os
mu_platform
uboot
mu_platform
  - MU_BASECORE
  - Common/MU
  - Silicon/ARM/MU_TIANO
  - Common/MU_TIANO
  - Common/MU_OEM_SAMPLE
```

We are using the `mu_platform` as workspace directory and it worked pretty well until update from `edk2-pytool-extensions==0.12.*` to `edk2-pytool-extensions~=0.23.0`. After the update, the `Edk2PlatformSetup().Invoke()` fails with following error:

```
SECTION - Init Self Describing Environment
SECTION - Loading Plugins
SECTION - Start Invocable Tool
PROGRESS - ## Resolving Git Submodule: MU_BASECORE
ERROR - Error when trying to resolve MU_BASECORE
ERROR - Invalid Git Repository at MU_BASECORE
SECTION - Summary
ERROR - Error
```
I did some debugging and it seems that the problem is in `submodule_resolve()` function being used with workspace path as repository root. In newer versions `GitPython` dependency was bumped and that deprecates searching a repo in parent repositories when calling `Repo()` constructor, [more here](https://github.com/gitpython-developers/GitPython/blob/main/git/repo/base.py#L184)

Does this really mean, that the workspace has to point to repository root folder? I find this a little bit creepy since it would force us to keep EDK-related stuff in repo root instead of in its own folder.

Thank you in advance!

**Comment by Javagedes** [comment]
hi @ljanik - I am incredibly sorry that I never saw this discussion. I do not get notified for discussion posts and will to resolve that.

I know it's been a year, but is this still an issue? Regardless I want to have an answer to anyone that sees this. `submodule_resolve` is not hardcoded to the root of the main repo, it is set to the workspace path, which is defined in your `PlatformBuild` (or equivalent)'s `GetWorkspaceRoot()` implementation. So this function could return `mu_platform` instead of the root of the repository itself.

If you absolutely need `GetWorkspaceRoot()` to be the root of the repository itself, then you would need to update the submodules inside your `.gitmodules` to be relative to your repository root rather than `mu_platforms`.

If I missed the mark and you had other issues, please let me know. I'd be happy to help (And I've updated my notifications for this repo to include discussions).

**Reply by ljanik** [reply]
Hi @Javagedes. Changing workspace path makes sense to me right now, but I think there was some side-effect that forced us to use different approach. I has been quite some time and to be honest, our contract ended and I can't even remember how we solved it.
Thank you anyway!
